encoding/encoding.go

```diff
@@ -0,0 +1,145 @@
+package encoding
+
+import (
+	"io"
+)
+
+// StringEncoder encodes string values.
+type StringEncoder interface {
+	// Encode a collection of strings.
+	// Callers should explicitly call `Reset` before
+	// subsequent call to `Encode`.
+	Encode(values StringIterator, writer io.Writer) error
+	// Reset should be called between `Encode` calls.
+	Reset()
+}
+
+// StringDecoder decodes int values.
+type StringDecoder interface {
+	// Decode the source bytes.
+	// Callers should explicitly call `Reset` before
+	// subsequent call to `Decode`.
+	Decode(src Reader) (StringIterator, error)
```

Comment:
I'm not sure the iterator can rewind during decoding though. Also doesn't look like we need that functionality. Thoughts?

Reply:
Also perhaps we just rename `StringDecoder` to `StringIterator`? Seems all it does it producing an iterator.

Reply:
It can't rewind during decoding. Let's just remove `StringDecoder` then. There already is a `StringIterator`.

Reply:
I meant currently the `StringIterator` has a `Rewind` method in the API due to embedding the `BaseIterator`, so if the iterator can't rewind during decoding, then the string decoder shouldn't return a `StringIterator`.

I was thinking perhaps we name the iterator passed to encoder a `RewindableStringIterator`, and the iterator returned by the decoder `ForwardIterator` taking inspiration from c++ std iterator naming patterns. IOW you'd have
```
type StringEncoder interface {
  Encode(it RewindableIterator) error
}

type ForwardStringIterator interface {
  Next() bool
  Current() string
  Err() error
  Close() error
}

type RewindableString Iterator {
  ForwardStringIterator

  Reset(values []string)
  Rewind()
}
```
Thoughts?